### PR TITLE
[Core] Update react form 

### PR DESCRIPTION
**Pull request by alisterdev**
- [x] Update to React form to support elements passed as a JSON object 
- [x] Added LorisElement as a generic element wrapper (used by React Form)
- [x] Abstracted button styles in props (for more flexibility)

>**Note about changes**: 
This update is the first step in converting Loris filter forms to use react (i.e dicom_archive style)

>**Note for Manual Testing:**
Make sure that Media, DICOM Archive, Issue Tracker and Candidate Parameters are all rendered as before

### CODE REVIEW

File: jsx/Form.js [hunk 1]
```diff
@@ -21,7 +28,9 @@ var FormElement = React.createClass({
     id: React.PropTypes.string,
     method: React.PropTypes.oneOf(['POST', 'GET']),
     class: React.PropTypes.string,
-    onSubmit: React.PropTypes.func
+    columns: React.PropTypes.number,
+    onSubmit: React.PropTypes.func,
+    onUserInput: React.PropTypes.func
```
Comment (jacobpenny) [inline jsx/Form.js]: Should add `formElements: React.PropTypes.array` I believe.

File: jsx/Form.js [hunk 2]
```diff
@@ -30,12 +39,56 @@ var FormElement = React.createClass({
       id: null,
       method: 'POST',
       class: 'form-horizontal',
+      columns: 1,
       fileUpload: false,
       onSubmit: function() {
         console.warn('onSubmit() callback is not set!');
+      },
+      onUserInput: function() {
+        console.warn("onUserInput() callback is not set!");
```
Comment (jacobpenny) [inline jsx/Form.js]: `formElements` should have a default of `[]` I think.

File: jsx/Form.js [hunk 3]
```diff
@@ -30,12 +39,56 @@ var FormElement = React.createClass({
...
+  getFormElements: function() {
+    var formElementsHTML = [];
+    var columns = this.props.columns;
+    var maxColumnSize = 12;
+    var colSize = Math.floor(maxColumnSize / columns);
+    var colClass = "col-xs-12 col-sm-" + colSize + " col-md-" + colSize;
+
+    // Render elements from JSON
+    var filter = this.props.formElements;
+    var userInput = this.props.onUserInput;
```
Comment (jacobpenny) [inline jsx/Form.js]: You could use `const` instead of `var` for all theses.

File: jsx/Form.js [hunk 4]
```diff
@@ -30,12 +39,56 @@ var FormElement = React.createClass({
...
+    if (filter) {
```
Comment (jacobpenny) [inline jsx/Form.js]: If we have a default of `[]` for `formElements` we can remove this conditional and use `map`.

```javascript
const formElementsHTML = this.props.formElements.map((element) => {
  return (
    <div className={colClass}>
      <LorisElement element={element} onUserInput={userInput} />
    </div>
  );
});
```

File: jsx/Form.js [hunk 5]
```diff
@@ -783,14 +850,58 @@ var ButtonElement = React.createClass({
...
   }
 });
+
+/**
+ * Generic form element.
+ */
+var LorisElement = React.createClass({
+
+  render: function() {
+    var props = this.props.element;
```
Comment (jacobpenny) [inline jsx/Form.js]: Can we call this `elementProps` instead of just `props`? Or just something to differentiate it from this component's `props`.

Comment (alisterdev) [inline jsx/Form.js]: Good point, will do

File: jsx/Form.js (hunk 4 above)
Comment (alisterdev) [inline jsx/Form.js]: I'm still an arrow function n00b but good point, this looks neater

**Comment by alisterdev** [comment]
Addressed all the comments to make cleaner code

**Pull request merged**